core/src/nonce/versioned.rs

```diff
@@ -0,0 +1,122 @@
+use core::mem::size_of;
+use defuse_borsh_utils::adapters::{BorshDeserializeAs, BorshSerializeAs};
+use hex_literal::hex;
+use near_sdk::borsh::{BorshDeserialize, BorshSerialize};
+use std::io::{self, Read};
+
+use crate::{
+    Nonce,
+    nonce::{expirable::ExpirableNonce, salted::SaltedNonce},
+};
+
+/// To distinguish between legacy nonces and versioned nonces
+/// we use a specific prefix individual for each version.
+/// Versioned nonce formats:
+/// - Legacy: plain `[u8; 32]`
+/// - VERSIONED: `VERSIONED_MAGIC_PREFIX (4 bytes) || VERSION (1 byte) || NONCE_BYTES (27 bytes)`:
+///     - V1: `SALT (4 bytes) || DEADLINE (8 bytes) || NONCE (15 random bytes)`
+#[derive(Clone, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+#[borsh(crate = "::near_sdk::borsh")]
+pub enum VersionedNonce {
+    Legacy(Nonce),
+    V1(SaltedNonce<ExpirableNonce<[u8; 15]>>),
+}
+
+// Allowed to maintain a clean API
+#[allow(clippy::fallible_impl_from)]
+impl From<Nonce> for VersionedNonce {
+    fn from(value: Nonce) -> Self {
+        MaybeVersionedNonce::deserialize_as(&mut value.as_ref()).unwrap_or_else(|_| unreachable!())
+    }
+}
+
+// Allowed to maintain a clean API
+#[allow(clippy::fallible_impl_from)]
+impl From<VersionedNonce> for Nonce {
+    fn from(value: VersionedNonce) -> Self {
+        const SIZE: usize = size_of::<Nonce>();
+        let mut result = [0u8; SIZE];
+
+        MaybeVersionedNonce::serialize_as(&value, &mut result.as_mut_slice())
+            .unwrap_or_else(|_| unreachable!());
+
+        result
+    }
+}
+
+struct MaybeVersionedNonce;
+
+impl MaybeVersionedNonce {
+    /// Magic prefixes (first 4 bytes of `sha256(<versioned_nonce>)`) used to mark versioned nonces:
+    pub const VERSIONED_MAGIC_PREFIX: [u8; 4] = hex!("5628f6c6");
+}
+
+impl BorshDeserializeAs<VersionedNonce> for MaybeVersionedNonce {
+    fn deserialize_as<R>(reader: &mut R) -> io::Result<VersionedNonce>
+    where
+        R: io::Read,
+    {
+        let mut prefix = [0u8; 4];
+        reader.read_exact(&mut prefix)?;
+
+        let versioned = if prefix == Self::VERSIONED_MAGIC_PREFIX {
+            VersionedNonce::deserialize_reader(reader)?
+        } else {
+            VersionedNonce::Legacy(Nonce::deserialize_reader(&mut prefix.chain(reader))?)
+        };
+
+        Ok(versioned)
```

Comment:
Nit: a bit cleaner to not use `?` and then wrap into `Ok` again IMO:
```suggestion
        if prefix == Self::VERSIONED_MAGIC_PREFIX {
            VersionedNonce::deserialize_reader(reader)
        } else {
            Nonce::deserialize_reader(&mut prefix.chain(reader)).map(VersionedNonce::Legacy)
        }
```